demos/doom/main.py

```diff
@@ -0,0 +1,155 @@
+from sysv_ipc import SharedMemory, ExistentialError
+import numpy as np
+from loguru import logger
+import cv2
+from demos.utils import get_all_from_queue
+from json import loads
+from os.path import exists
+from os import environ
+from subprocess import Popen
+
+
+DOOM_VIDEO_ID = 666
+DOOM_INPUT_ID = 667
+DOOM_OUTPUT_ID = 668
+
+SCREENWIDTH = 320
+SCREENHEIGHT = 200
+NUM_COLS = 48
+
+
+class Doom:
+    """This is an interactive demo that adapts the Doom game for the SSS by capturing the game's video output"""
+
+    demo_time = 50  # None for a game
+
+    # User input is passed through input_queue
+    # Game output is passed through output_queue
+    # Screen updates are done through the screen object
+    def __init__(self, input_queue, output_queue, screen):
+        # Provide the framerate in frames/seconds and the amount of time of the demo in seconds
+        self.frame_rate = 50
+
+        self.input_queue = input_queue
+        self.output_queue = output_queue
+        self.screen = screen
+
+        self.game_installed = exists("demos/doom/chocolate-doom")
+
+        self.shared_mem_init = False
+
+        # Initialize game and allocate shared memory
+        if self.game_installed:
+
+            environ["SDL_VIDEODRIVER"] = "dummy"
+            Popen(["./demos/doom/chocolate-doom", "-iwad", "assets/miniwad.wad"])
+
+            # init a connection to shared memory locations here
+            try:
+                self.shm = SharedMemory(DOOM_VIDEO_ID, 0, 0)
+                self.shm.attach(0, 0)
+
+                self.shm_input = SharedMemory(DOOM_INPUT_ID, 0, 0)
+                self.shm_input.attach(0, 0)
+
+                self.shm_output = SharedMemory(DOOM_OUTPUT_ID, 0, 0)
+                self.shm_output.attach(0, 0)
+
+                self.shared_mem_init = True
+            except (ExistentialError, Exception) as e:
+                logger.error("Could not establish connection with shared memory")
+                logger.error(e)
+                exit(0)
+
+        # Color map for the SSS
+        self.num_to_pixel = {
+            0: 0x0,
+            1: 0x0,
+            2: 0x0,
+            3: 0x1,
+            4: 0x1,
+            5: 0x2,
+            6: 0x2,
+            7: 0x2,
+            8: 0x2,
+            9: 0xA,
+            10: 0xE,
+            11: 0xF,
+            12: 0xF,
+        }
+
+        self.screen_max = 0
+        self.screen_min = 0
+
+        # Init numpy array for screen
+        self.arr = []
+        for i in range(NUM_COLS):
+            self.arr.append([])
+            for j in range(NUM_COLS):
+                self.arr[i].append(0)
+
+    def run(self):
+        # Create generator here
+        while self.game_installed and self.shared_mem_init:
```

Comment:
Just a style suggestion and you can totally ignore it. I would suggest you don't make this a while condition since it is not going to ever change. I would just do an if statement to check if things were not set up correctly, that way if it gets past the if statement, then you know everything is good to go (like a [guard statement](https://medium.com/@scadge/if-statements-design-guard-clauses-might-be-all-you-need-67219a1a981a)). This is what I'm thinking:


```python
if not self.game_installed or not self.shared_mem_init:
    self.screen.draw_text(
          self.screen.x_width // 2 - 10,
          self.screen.y_height // 2 - 4,
          "ERROR INITIALIZING DOOM",
          push=True,
      )
      while True:
          yield

# [...code for normal execution...]
```